### DISCUSSION TITLE
getting_started/step_by_step/scripting_player_input

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#34] [User-contributed notes]
# getting_started/step_by_step/scripting_player_input

Building upon the previous lesson, Creating your first script, let's look at another important feature of any game: giving control to the player. To add this, we need to modify our sprite_2d.gd cod...

https://docs.godotengine.org/en/latest/getting_started/step_by_step/scripting_player_input.html

**Comment by Stephen-Safee** [comment]
This is good, but how do I code controls for a mobile device? So, no keyboard or mouse.

**Reply by Calinou** [reply]
Use the TouchScreenButton node to create touch-based controls, and use the relevant Input singleton methods to read the accelerometer/gyroscope.

This is not showcased in this step-by-step demo to keep the code as short as possible, but perhaps it should be done as an optional addendum for those following the tutorial on the Android editor.

**Reply by Stephen-Safee** [reply]
> Use the TouchScreenButton node to create touch-based controls, and use the relevant Input singleton methods to read the accelerometer/gyroscope.
> 
> This is not showcased in this step-by-step demo to keep the code as short as possible, but perhaps it should be done as an optional addendum for those following the tutorial on the Android editor.

Thanks. Is there a guide anywhere at all for this? That is, for creating mobile controls using the TouchScreenButton node or something else?

**Reply by Calinou** [reply]
> Is there a guide anywhere at all for this?

Not that I know of, but some of the official demos like [Platformer 3D](https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer) feature touch controls that automatically appear on mobile devices.

The process should be straightforward:

- Create input actions like you'd do to support keyboard and mouse.
- Create TouchScreenButton nodes with their `action` property set to match the name of the input action they should trigger when pressed.
- Test on a mobile device. TouchScreenButtons can be configured to appear on touchscreens only, or at all times (for testing purposes), but testing on a mobile device is preferred to ensure it actually plays well.

**Reply by glumdrop** [reply]
iOS / Xogot Note – Using Virtual Controller for Touch Input

If you’re running Godot on iOS (via Xogot) and want to use touch screen controls, you can map movement actions to the virtual controller instead of keyboard input:

	1.	Open Project Settings → scroll to Input Devices → enable Virtual Controller.
This adds a touch screen controller overlay when the game runs.

	2.	Go to Input Map (under General) and add the following actions individually: move_right, move_left, move_up.

	3.	For each action, under Action Event, select Joypad Axes and assign these inputs:
	•	Move Right → Joypad Axis 0 + (Left Stick Right, Joystick 0 Right)
	•	Move Left → Joypad Axis 0 + (Left Stick Left, Joystick 0 Left)
	•	Move Up → Joypad Axis 0 + (Left Stick Up, Joystick 0 Up)

	4.	Once mapped, replace the tutorial’s "ui_left", "ui_right", "ui_up" references with your new actions (move_left, move_right, move_up) in the script.

This allows you to follow the tutorial code while supporting touch controls seamlessly.

Tutorial on Input Maps can be found here
https://docs.godotengine.org/en/4.4/getting_started/first_2d_game/03.coding_the_player.html

**Reply by Obsessive-Coder** [reply]
There is no option for enabling virtual controller lmao why are there so many lies and false instructions when it comes dev tutorials hahahahhahaha. So sick if it

**Reply by 1337cookie** [reply]
> There is no option for enabling virtual controller lmao why are there so many lies and false instructions when it comes dev tutorials 

Glumdrop above is talking about "Xogot" which is a paid app built from the Godot source for touch screen/UI compatibility on Apple mobile devices. Xogot implements touch controls using a "Virtual Joystick" addon but doesn't mention which one. There's currently 3 addons called "Virtual Joystick" on the AssetLib, it's probably one of those.

**Comment by dongb5** [comment]
emmmm Weird coordinate system. The X axis points to right, and the Y axis points to down, which is fine that same to the image coordinate. But the zero degree points to UP. If I use cos() / sin() to calculate the component of velocity along X and Y, it is like:
```
position.x += speed * delta * sin(rotation)
position.y += - speed * delta * cos(rotation)
```
instead of vx = speed * cos(rotation) and vy = speed * sin(rotation). Namely, why the rotation coordinate is different from the X-Y coordinate?

**Reply by lemimes** [reply]
Using the [GDScript training module](https://gdquest.github.io/learn-gdscript/?ref=godot-docs) you will learn that the game screen uses coordinates in a strange way, but it makes sense once they explain it to you (has to do with the way pixels are rendered on a screen.)

**Comment by Keiko-Ndev** [comment]
I manage to ultimately get the same results with my code by defining a new variable called move that would modify the velocity calculation, basically skipping the last step since I wanted to experiment by myself. If somebody could explain the difference between using the Vector2.ZERO method and this one it would be appreciated. Other than that I am leaving this for reference for anybody curious:

```gdscript
extends Sprite2D
var speed = 500
var angular_speed = PI
func _process(delta):
	var direction = 0
	if Input.is_action_pressed("ui_left"):
		direction = -1
	if Input.is_action_pressed("ui_right"):
		direction = 1
	rotation += angular_speed * direction * delta
	var move = 0
	if Input.is_action_pressed("ui_up"):
		move = 1
	if Input.is_action_pressed("ui_down"):
		move = -1
	var velocity = Vector2.UP.rotated(rotation) * speed * move
	position += velocity * delta
```

**Reply by shiverczar** [reply]
Originally wrote up my understanding w/o looking up info, decided to look stuff up to be a bit more informative.

"position" is actually a Vector2 that provides X/Y coordinates for the location. (control + click on position to see more detail; it's how I know this xD )

var velocity = Vector2.ZERO => velocity = (0,0)

If we do not press the up arrow, velocity remains (0,0) for this cycle through the _process function, and when we get to the position change:

position += velocity * delta => assuming velocity remains as 0,0 [i.e. no input], 0 * delta = 0 for both X and Y values. 'position' is not changed.

^ This is how the Vector2.ZERO applies to the 'base' version of the code.

Your version, if the up / down arrows are not pressed, includes a multiplier by 0 in the declaration of the vector. 
var velocity = Vector2.UP.rotated(rotation) * speed * move
=> (X, Y) * 500 * 0 => (0,0) -- X and Y are irrelevant since they get multiplied by zero (move).

Base version:
	if Input.is_action_pressed("ui_up"):
		velocity = Vector2.UP.rotated(rotation) * speed

Your version just sets move = 1 instead, and then multiplies that 1 by (essentially) the same velocity assignment. (X,Y) * speed * 1 = (X,Y) * speed, so it ends up with the same result.


TLDR:
Your version creates the variable 'move' which, when not receiving input, zeroes out the velocity declaration math resulting in a (0,0) vector.

Base version just starts with a (0,0) vector and only does anything else if there's input.

**Comment by jeremy-jarvis** [comment]
The [Moving When Pressing Up](https://docs.godotengine.org/en/stable/getting_started/step_by_step/scripting_player_input.html#moving-when-pressing-up) step does not instruct the reader to un-comment this line, but it must be un-commented in order for the code to work.

position += velocity * delta

**Comment by Zootz** [comment]
My left and up is slow and I have to tap it to move, while my right can keep on turning as long as I hold it down. How do I fix this?

**Reply by n-hdev** [reply]
Sounds like you used is_action_just_pressed() rather than is_action_pressed() function for the left key. Correct your code to use the is_action_pressed() function.

**Reply by Zootz** [reply]
Thanks so much. It worked! Must've been typing faster than the wind.

**Comment by ArztVielfrass** [comment]
Is there a way to copy and paste code in the editor and indent it on that line? I keep having to fix indentation every time I paste code. I'm aware that I can: paste code, highlight that code, right click, then indent. I was hoping that it does it automatically.

**Comment by Falco-Boehnke** [comment]
I am missing one piece for my understanding here:

Is Vector2.UP in this script a local Vector based on the node, or is it a global Vector2.UP, based on the editor, but then rotated to fit the current nodes rotation using a member variable that is always based on the instance?

**Comment by asadovkamran** [comment]
hm, for some reason i can rotate right but can't rotate left...

**Reply by jalanglois94-boop** [reply]
I know this is older but for anyone looking now, this happened to me and it was because my "rotation += angular...." was only indented under my UI-Right instead of not being indented. 

"Your rotation += ... line is indented under the if Input.is_action_pressed("ui_right"): block. This means it only runs if the right key is pressed, and does nothing if you press left. So when you press left, direction = -1, but the rotation line is never executed. "

Hope that helps!

**Comment by Doeryo45** [comment]
Hi guys, somehow this program is not working. Godot icon on screen is stuck in its place and not moving with _process function, I'm using version 4.4.1 btw

**Reply by TurirBarym** [reply]
have you re-added the "position += velocity * delta" line at the end of your code?
It is part of the complete script example, but not mentioned in the text that it needs to be re-added/ uncommented

**Reply by Joe2500** [reply]
Make sure to click the "input" button in the test application

**Comment by trev33b** [comment]
This example doesn't properly handle the case when both the left and right arrows are pressed simultaneously.  Instead of cancelling each other out, the Right Arrow gets precedence since it's checked last.  I fixed this by using the following code:

	var direction = 0
	if Input.is_action_pressed("ui_left"):
		direction -= 1
	if Input.is_action_pressed("ui_right"):
		direction += 1

With my code, I decrement & increment the direction instead of explicitly setting it to 1 or -1.  This allows simultaneous presses to cancel each other out.  The same could be done with the move variable looking at Up & Down keys suggested by @Keiko-Ndev in these notes.

**Reply by Calinou** [reply]
Note that there is no universally correct approach to this problem, as the current behavior may be preferred in some cases (although I consider it less ideal than last-input priority). See also https://github.com/godotengine/godot-proposals/issues/12235.

**Reply by trev33b** [reply]
Fair enough.   You're right in that it really depends on what you're trying to achieve.  It just seemed that in this example the input priority shouldn't be opinionated since we're just playing around at this point to see what one can do and I thought seeing the ability to handle multiple simultaneous key presses might be eye opening to new developers who might care about input for action games, especially since my code change to support this was trivial and takes the same number of lines of code.

**Reply by VoxelHat** [reply]
How about this:

if Input.is_action_pressed("ui_left") && !Input.is_action_pressed("ui_right"):
		direction = -1
elif Input.is_action_pressed("ui_right") && !Input.is_action_pressed("ui_left"):
		direction = 1

When I press one key, the object rotates; pressing both keys stops it.

**Comment by Explosivevortex** [comment]
Uhh I got a weird issue with this code and the one from the previous lesson.  I've typed everything right, but instead of doing what it's supposed to, the icon moves in a clockwise/counter-clockwise circle whenever I press the left or right arrow key, and does nothing when I press the up arrow key.  

In the previous lesson the code also moved in this way when I typed in the code to make it just rotate.  Anybody know what the heck is going on?

**Reply by Explosivevortex** [reply]
Nevermind, fixed it by starting a new project, that was weird.

**Reply by ReskiNN** [reply]
Lol i got the same problem but starting a new project does nothing, any ideas?

**Comment by JiaoMkII** [comment]
I followed the tutorial to the letter, copied the code even, and I am still having a problem.

When I press left or right my sprite starts to move both forward and along the angle of PI in the respective direction. It does not spin in place and I can not figure out why.

**Comment by FrostBite0640** [comment]
I have put every line perfectly with no errors but the player wont move in debug mode

**Reply by Falco-Boehnke** [reply]
Can you show your script?

**Reply by LanaMarieO3** [reply]
I'm having the same issue. I copied the script they provided, and my "character" STILL won't move.

**Reply by netfab** [reply]
I just had the exact same problem. To fix it I had to :
- save everything (scene and script)
- close the Godot editor
- restart everything

And TADAAAA : it is moving on keys press.
Probably a cache problem.

**Comment by IordanisKokk** [comment]
So, I don't know if anyone would find this useful but, I am also following this guide to try to learn Godot and Game dev properly this time. 

After completing the script, I though that why not try to make an 8-directional movement script rather than the tank controls that the guide suggest. So, if anyone is interested, here is the complete script for that.

```gdscript
extends Sprite2D

var speed = 400
var angular_speed = PI


func _process(delta):
	var direction = Vector2.ZERO
	
	if Input.is_action_pressed("ui_left"):
		direction += Vector2.LEFT
	if Input.is_action_pressed("ui_right"):
		direction += Vector2.RIGHT
	if Input.is_action_pressed("ui_up"):
		direction += Vector2.UP
	if Input.is_action_pressed("ui_down"):
		direction += Vector2.DOWN
		
	position += speed * direction * delta
```

Figuring this out actually made me smile haha.

**Comment by whuffman36** [comment]
Thanks for the great tutorial! Just out of curiosity, why are the different input methods (ie `"ui_right"`, `"ui_left"`, etc) accessed as strings rather than something like an enum? I have been bit by string accessors in the past as it can be easy to misspell and debug.

It was mentioned in another comment that GDScript does not really have globals in the sense that other languages do. Is there a way to define our own global enum mapping (or any other class for that matter) file that can be included or imported in other files? For example:

```gdscript
# enum.gd script

enum UiMethod {
    UI_RIGHT,
    UI_LEFT,
    ...
}

func UiMethodToString(method: UiMethod) -> String:
    ...
```

```gdscript
# sprite_2d.gd script

extends Sprite2D

# Does something like this exist ?
import enum.gd
...
```

**Reply by Calinou** [reply]
> Thanks for the great tutorial! Just out of curiosity, why are the different input methods (ie "ui_right", "ui_left", etc) accessed as strings rather than something like an enum? I have been bit by string accessors in the past as it can be easy to misspell and debug.

It's possible to use constants instead, but you need to write these manually as Godot doesn't automatically generate constants based on input names.

> Is there a way to define our own global enum mapping (or any other class for that matter) file that can be included or imported in other files

Yes, using an [autoload](https://docs.godotengine.org/en/stable/tutorials/scripting/singletons_autoload.html) that has Singleton enabled (which is the default). This way, it can be accessed globally using its name without needing to `preload()` it into a constant beforehand.

**Comment by admin** [comment]
*Originally posted by [madaomax](https://github.com/madaomax):*

> When commenting out these two lines:
> ```
> #var velocity = Vector2.UP.rotated(rotation) * speed
> #position += velocity * delta
> ``` 
> Should comment out this line too:
> ```
> #rotation += angular_speed * delta
> ```

**Comment by admin** [comment]
*Originally posted by [ryddem](https://github.com/ryddem):*

> you could also add an input so ur sprite moves backwards
> 
> var velocity = Vector2.ZERO
> if Input.is_action_pressed("ui_down"):
> velocity = Vector2.DOWN.rotated(rotation) * speed
> position += velocity * delta

**Reply by admin** [reply]
*Originally posted by [ryddem](https://github.com/ryddem):*

> if u don't want to type all of this manual copy ur forward input which is up and change the words to down especially in
> 
> velocity = Vector2.UP.rotated(rotation) * speed
> 
> make sure when replacing up with down it has to be capitalized
